service/resource/endpoint/delete.go

```diff
@@ -11,60 +10,72 @@ import (
 )
 
 func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteState interface{}) error {
-	endpointToApply, err := toEndpoint(deleteState)
+	if deleteState == nil {
+		return nil // Nothing to do.
+	}
+
+	k8sEndpoint, err := toK8sEndpoint(deleteState)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	if reflect.DeepEqual(endpointToApply, Endpoint{}) {
+	if !isEmptyEndpoint(*k8sEndpoint) {
 		return nil
 	}
 
-	k8sAddresses := []apiv1.EndpointAddress{}
-	for _, endpointIP := range endpointToApply.IPs {
-		k8sAddress := apiv1.EndpointAddress{
-			IP: endpointIP,
-		}
-		k8sAddresses = append(k8sAddresses, k8sAddress)
+	err = r.k8sClient.CoreV1().Endpoints(k8sEndpoint.Namespace).Delete(k8sEndpoint.Name, &metav1.DeleteOptions{})
+	if err != nil {
+		return microerror.Mask(err)
 	}
 
-	if len(k8sAddresses) == 0 {
-		err = r.k8sClient.CoreV1().Endpoints(endpointToApply.ServiceNamespace).Delete(endpointToApply.ServiceName, &metav1.DeleteOptions{})
-		if err != nil {
-			return microerror.Mask(err)
-		}
-		return nil
-	}
+	return nil
+}
 
-	k8sEndpoint, err := r.k8sClient.CoreV1().Endpoints(endpointToApply.ServiceNamespace).Get(endpointToApply.ServiceName, metav1.GetOptions{})
+func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	deleteState, err := r.newDeleteChangeForDeletePatch(ctx, obj, currentState, desiredState)
 	if err != nil {
-		return microerror.Mask(err)
-	}
-	for i := range k8sEndpoint.Subsets {
-		k8sEndpoint.Subsets[i].Addresses = k8sAddresses
+		return nil, microerror.Mask(err)
 	}
-
-	_, err = r.k8sClient.CoreV1().Endpoints(endpointToApply.ServiceNamespace).Update(k8sEndpoint)
+	updateState, err := r.newDeleteChangeForUpdatePatch(ctx, obj, currentState, desiredState)
 	if err != nil {
-		return microerror.Mask(err)
+		return nil, microerror.Mask(err)
 	}
+	patch := framework.NewPatch()
+	patch.SetDeleteChange(deleteState)
+	patch.SetUpdateChange(updateState)
 
-	return nil
+	return patch, nil
 }
 
-func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
-	delete, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
+func (r *Resource) newDeleteChangeForDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*apiv1.Endpoints, error) {
+	desiredEndpoint, err := toEndpoint(desiredState)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
-	patch := framework.NewPatch()
-	patch.SetDeleteChange(delete)
+	currentEndpoint, err := toEndpoint(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
 
-	return patch, nil
+	deleteState := Endpoint{
```

Comment:
So this is a value which is returned but at the same time you return nil in error cases. That is inconsistent. We should always return the zero value in error cases. I would suggest to deal with `*Endpoint` and keep `nil`.

Reply:
This was due to me slipping up with variable naming. The actual return value is `*apiv1.Endpoints`. I fixed the naming, this should be correct be as you intended.